README.md
Spike for PHP
=============

[![Build Status](https://travis-ci.org/issei-m/spike-php.svg)](https://travis-ci.org/issei-m/spike-php)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/issei-m/spike-php/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/issei-m/spike-php/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/issei-m/spike-php/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/issei-m/spike-php/?branch=master)
[![License](https://poser.pugx.org/issei-m/spike-php/license.svg)](https://packagist.org/packages/issei-m/spike-php)

Latest release: [v1.0-BETA5](https://packagist.org/packages/issei-m/spike-php#v1.0-BETA5)

The client of https://spike.cc REST api for PHP (5.4+, 7+, HHVM).

Basic Usage
-----------

First, initialize the `Spike` object with your **api secret key**. It's the entry point for accessing the all api interfaces:

```php
$spike = new \Issei\Spike\Spike('your_api_secret_key');
```

### Create a new charge

To create a new charge, you have to build a `ChargeRequest` object. It can be specified `card token`, `amount`, `currency` and some related products. Next, call `charge()` method with it. If charge succeeded this method will return the new `Charge` object generated by REST api:

```php
// The token's id is retrieved by SPIKE Checkout (client side script) usually.
$token = new \Issei\Spike\Model\Token('tok_xxxxxxxxxxxxxxxxxxxxxxxx');

$request = new \Issei\Spike\ChargeRequest();
$request
    ->setToken($token)
    ->setAmount(123.45, 'USD')
    ->setCapture(true) // If you set false, you can delay capturing.
;

$product = new \Issei\Spike\Model\Product('my-product-00001');
$product
    ->setTitle('Product Name')
    ->setDescription('Description of Product.')
    ->setPrice(123.45, 'USD')
    ->setLanguage('EN')
    ->setCount(3)
    ->setStock(97)
;

// The product can be added any times.
$request->addProduct($product);

/** @var $createdCharge \Issei\Spike\Model\Charge */
$createdCharge = $spike->charge($request);
```

Tips: You can pass the `Token`'s id directly instead of generating the new `Token` object:

```php
$request->setToken('tok_xxxxxxxxxxxxxxxxxxxxxxxx');
```

**NOTE**: If you want to know how to get a `card token`, read [Request a token](#request-a-token) section.

### Find a charge

Call `getCharge()` method with charge id:

```php
/** @var $charge \Issei\Spike\Model\Charge */
$charge = $spike->getCharge('20150101-100000-xxxxxxxxxx');
```

### Capture the charge

If you have a charge which has not been captured, you can use `capture()` method to capture it:

```php
/** @var $charge \Issei\Spike\Model\Charge */
$capturedCharge = $spike->capture($charge);
```

### Refund the charge

Call `refund()` method with the `Charge` object that you want to refund:

```php
/** @var $charge \Issei\Spike\Model\Charge */
$refundedCharge = $spike->refund($charge);
```

Tips: You can pass the `Charge`'s id directly instead of generating/retrieving the `Charge` object:

```php
$refundedCharge = $spike->refund('20150101-100000-xxxxxxxxxx');
```

### Retrieve the all charges

Call `getCharges()` method. it returns an array containing the `Charge` objects.

```php
/** @var $charges \Issei\Spike\Model\Charge[] */
$charges = $spike->getCharges();
```

#### Paging

You can specify the limit of number of records at 1st argument (10 records by default):

```php
$charges = $spike->getCharges(5);
```

If you pass a `Charge` object (or ID as a string directly) into 2nd argument, you can retrieve charges that older than (passed charge is NOT included to list):

```php
$nextCharges = $spike->getCharges(5, $charges[count($charges) - 1]);
```

At 3rd argument, you can also specify the charge (or ID as a string directly) object if you want to retrieve charges that newer than (passed charge is NOT included to list):

```php
$nextCharges = $spike->getCharges(5, $charges[count($charges) - 1], ...);
```

### Request a token

If you have contracted with https://spike.cc to request a new token, you can get a new token by `requestToken()` method with `TokenRequest`:

```php
$request = new \Issei\Spike\TokenRequest();
$request
    ->setCardNumber('4444333322221111')
    ->setExpirationMonth(12)
    ->setExpirationYear(19)
    ->setHolderName('Taro Spike')
    ->setSecurityCode('123')
    ->setCurrency('JPY')
    ->setEmail('test@example.jp')
;

/** @var $charge \Issei\Spike\Model\Token */
$token = $spike->requestToken($request);
```

Of course, you can create a new charge with it:

```php
$request = new \Issei\Spike\ChargeRequest();
$request
    ->setToken($token)
    // ...
;

/** @var $charge \Issei\Spike\Model\Charge */
$charge = $spike->charge($request);
```

### Find a token

Call `getToken()` method with token id:

```php
/** @var $token \Issei\Spike\Model\Token */
$token = $spike->getToken('tok_xxxxxxxxxxxxxxxxxxxxxxxx');
```

Installation
------------

Use [Composer] to install the package:

```
$ composer require issei-m/spike-php
```

Contributing
------------

1. Fork it
2. Create your feature branch
3. Commit your change and push it
4. Create a new pull request

[SPIKE Checkout]: https://spike.cc/dashboard/developer/docs/references#a1
[Composer]: https://getcomposer.org


CHANGELOG.md
Changelog
=========

- 1.0-BETA5 (201x-xx-xx)
  - Start supporting PHP 7.2
  - Ensure dev stuff (e.g. tests, .gitignore, etc) is removed when package installation by adding `.gitattributes`

- 1.0-BETA4 (2016-12-22)
  - Starting PHP 7.1 support
  - Fix wrong parameter of charge api
  - Rename ChargeRequest::getCard(), setCard() methods to getToken(), setToken() (BC break)
  - Rename RecursiveObjectFactoryConverterBuilder::getBuilder() method to getConverter() (BC break)

- 1.0-BETA3 (2015-05-14)
  - Spike::getCharges(), ::setCard() and ::refund() methods now allow to be passed related object as a string
  - added ObjectList model object
  - now supports Dispute object
  - now supports capturing the charge
  - renamed Spike::getToken() method to requestToken() (BC Break)
  - Spike::getToken() method is now used to retrieve the token object

- 1.0-BETA2 (2015-03-30)
  - Charge/RefundFactory now uses system default timezone if none is specified
  - Spike::charge() now avoids error even if request has not set any value at all
  - added Spike::getToken to support retrieving a token interface (fixed #2)
  - added ObjectConverterInterface and replaced Spike's 2nd constructor argument with it (BC Break)
  - moved MoneyFactoryTrait to Util (BC Break)
  - added DateTimeUtil in favor of using TimeZoneSpecifiableTrait (BC Break)
  - Charge/Token now implement __toString()

- 1.0-BETA1 (2015-01-16)
  - renamed ApiErrorException to RequestException and changed its constructor signature
  - now uses Money class for dealing with money
  - Exception now extends \RuntimeException
  - now supports GuzzleHttp client as implementation of ClientInterface
  - changed Spike::__construct signature

- 1.0-alpha1 (2015-01-08)
  - first alpha release


res/example/index.php
<?php

require __DIR__ . '/../../vendor/autoload.php';
require __DIR__ . '/definition.php';

$spike = new \Issei\Spike\Spike(API_SECRET);

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

echo '<h1>Retrieve the token</h1>';

$req = new \Issei\Spike\TokenRequest();
$req
    ->setCardNumber('4444333322221111')
    ->setExpirationMonth(12)
    ->setExpirationYear(19)
    ->setHolderName('Taro Spike')
    ->setSecurityCode('123')
    ->setCurrency('JPY')
    ->setEmail('test@example.jp')
;
dump($token = $spike->requestToken($req));

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

echo '<h1>New charge</h1>';

$req = new \Issei\Spike\ChargeRequest();
$req
    ->setAmount(100, 'JPY')
    ->setToken($token)
    ->addProduct(
        (new \Issei\Spike\Model\Product(uniqid('product-', true)))
            ->setTitle('Title')
            ->setDescription('Description')
            ->setCount(1)
            ->setPrice(100, 'USD')
            ->setLanguage('JA')
    )
;
dump($spike->charge($req));

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

echo '<h1>Retrieve charges</h1>';

dump($charges = $spike->getCharges(5));

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

echo '<h1>Retrieve a charge</h1>';

dump($charge = $spike->getCharge($charges[0]));

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

echo '<h1>Refund the charge</h1>';

dump($spike->refund($charge));
